presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,244 @@
+            boolean nullAllowed = false;
+            for (Page page : buildPages) {
+                Block block = page.getBlock(channel);
+                for (int i = 0; i < block.getPositionCount(); i++) {
+                    Object value = TypeUtils.readNativeValue(descriptor.buildType, block, i);
+                    if (value == null) {
+                        nullAllowed = true;
+                    }
+                    else {
+                        values.add(value);
+                    }
+                }
+            }
+            return Domain.create(ValueSet.copyOf(descriptor.buildType, values.build()), nullAllowed);
+        }
+    }
+
+    public static class Descriptor
+    {
+        private final Type buildType;
+        private final int buildChannel;
+        private final ColumnHandle probeColumn;
+
+        public Descriptor(Type buildType, int buildChannel, ColumnHandle probeColumn)
+        {
+            this.buildType = buildType;
+            this.buildChannel = buildChannel;
+            this.probeColumn = probeColumn;
+        }
+    }
+
+    private final List<Descriptor> descriptors;
+    private final ImmutableList.Builder<CompletableFuture<TupleDomain<ColumnHandle>>> partitions;
+    private final CompletableFuture<CompletableFuture<Optional<TupleDomain<ColumnHandle>>>> result;
+
+    public DynamicFilter(List<Descriptor> descriptors)
+    {
+        this.descriptors = descriptors;
+        this.partitions = new ImmutableList.Builder<>();
+        this.result = new CompletableFuture<>();
+    }
+
+    /// Decorate existing page source provider, to allow dynamic filtering.
+    public PageSourceProvider wrap(PageSourceProvider pageSourceProvider)
+    {
+        return (Session session, Split split, TableHandle table, List<ColumnHandle> columns) -> {
+            Optional<TupleDomain<ColumnHandle>> dynamicPredicate = result()
+                    .getNow(Optional.empty()); // If dynamic filter is not available, don't block execution (to prevent deadlock with other operators).
+            TableHandle filteredTable = dynamicPredicate
+                    .map(predicate -> pageSourceProvider.applyDynamicFilter(session, table, predicate))
+                    .orElse(table);
+            return pageSourceProvider.createPageSource(session, split, filteredTable, columns);
+            // TODO: implement applyDynamicFilter() method for the anonymous class (to allow recursive dynamic filtering).
+        };
+    }
+
+    public Collector collector()
+    {
+        return new Collector();
+    }
+
+    /// Each CollectOperator should submit its own predicate into a separate future (one for each partition).
+    public void addFuture(CompletableFuture<TupleDomain<ColumnHandle>> future)
+    {
+        partitions.add(future);
+    }
+
+    /// Must be called after all futures have been added, i.e. by CollectOperatorFactory::noMoreOperators() method.
+    public void noMoreFutures()
```

Comment:
This shouldn't be necessary, because the number of futures is know at planning time.
It's `buildContext.getDriverInstanceCount().orElse(1);`  (see `createLookupSourceFactory`)

Reply:
Thanks, done at 39a2eb47da.